# My first Data Science Project

**My aim is to:**

    .collect and analyze data about mobile apps available on Google Play and the App Store.




**In following code we will:**
    
    >import reader from csv module
    >then after open the corresponding file by using open() function.
    >and after opening ,convert that file in list data-type by using list() function
    >Assign header row of each data set to a variable by using list indexing.
    >then print some row to check and verify.

In [1]:
from csv import reader
#open google play store data
opened_file=open('googleplaystore.csv',encoding='utf-8')
read_file=reader(opened_file)
Android=list(read_file)
Android_header=Android[0]
Android=Android[1:]

#Now open Apple store data
opened_file=open('Applestore.csv',encoding='utf-8')
read_file=reader(opened_file)
Ios=list(read_file)
Ios_header=Ios[0]
Ios=Ios[1:]

print(Android[:3],"\n\n",Ios[:3])

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']] 

 [['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']]


**In following code :**

    >We'll start by opening and exploring these two data sets.
    >To make them easier  to explore, we created a function named explore_data().
    >By using explore_data() we can repeatedly use to print rows in a readable way.

In [2]:
def explore_data(dataset,start ,end, rows_and_columns=False):
    dataset_slice=dataset[start:end]
    for row in dataset_slice:
        print(row)
        print("\n")#Add a empty line betwee rows.
    if rows_and_columns:
        print("No. of rows:",len(dataset))
        print("No. of columns:",len(dataset[0]))
print(Android_header)
print('\n')
explore_data(Android,0,3,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


No. of rows: 10841
No. of columns: 13


**In following code:**

    >we'll print header row of Ios.
    >And thenafter call explore_data function for print some row of  Ios.
    >No. of rows can be selected by sending start and end value of dataset.

In [3]:
print(Ios_header)
print('\n')
explore_data(Ios,0,3,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


No. of rows: 7197
No. of columns: 16


# Data cleaning (Delete unwanted and wrong data.).
**Before beginning our analysis, we need to make sure the data we analyze is accurate, otherwise the results of our analysis will be wrong. This means that we need to:**

    >>Detect inaccurate data, and correct or remove it.
    >>Detect duplicate data, and remove the duplicates.
    >>Recall that at our company, we only build apps that are free to download and install, and that are directed toward an 'English-speaking audience.'

**This means that we'll need to:**

    >>Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
    >>Remove apps that aren't free.
    
**This process of preparing our data for analysis is called data cleaning.**

In Google Play dataset  we can see that one of the discussions outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [4]:
print(Android[10472])
print("\n")
print(Android_header)
print("\n")
print(Android[0])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


**We can see that in rating section of 'Life Made WI-Fi Touchscreen Photo Frame', rating is '19' . Since maximum rating for google play store is '5' , that means there is error in this row , so in following code:**

    >>Delete that  row
    >>For making sure that row has been deleted we'll print length of dataset before and after deletion .
    >>Make sure that following code should be run only once ,otherwise this will delete other important data. For that we'll comment delete line after once it'll we run.

In [5]:
print(len(Android))
del(Android[10472])
print(len(Android))

10841
10840


# Removing Duplicate data:

### PART 1:

**If we explaore google play data_set ,  will find that there are some entries which are more than once. Before processing any operation we have to remove those duplicate data.:**

      >>for instance entries for instagram is four times.

In [6]:
for apps in Android:
    App_name=apps[0]
    if  App_name=="Instagram":
        print(apps)
        print("\n")

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




**In above code we seen that there are multiple entries for Instagram, Since we can't notice each and every data, we have to find all duplicate entries.**

In [7]:
duplicate_apps=[]
unique_apps=[]
for apps in Android:
    apps_name=apps[0]
    if  apps_name in unique_apps:
        duplicate_apps.append(apps_name)
    else :
        unique_apps.append(apps_name)

print("No. of duplicate data : ",len(duplicate_apps))
print("\nExample of duplicate data:",duplicate_apps[:25])

No. of duplicate data :  1181

Example of duplicate data: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho', 'join.me - Simple Meetings', 'Invoice 2go — Professional Invoices and Estimates', 'SignEasy | Sign and Fill PDF and other Documents']


**We don't want to count certain apps more than once when we analyze data, so we need to remove the duplicate entries and keep only one entry per app. One thing we could do is remove the duplicate rows randomly, but we could probably find a better way.**

**If you examine the rows we printed two cells above for the Instagram app, the main difference happens on the fourth position of each row, which corresponds to the number of reviews. The different numbers show that the data was collected at different times. We can use this to build a criterion for keeping rows. We won't remove rows randomly, but rather we'll keep the rows that have the highest number of reviews because the higher the number of reviews, the more reliable the ratings.**

**To do that, we will:**

    >>Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
    >>Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

## PART 2

**we'll start by creating a dictionary reviews_max={}**

In [8]:
reviews_max={}
for apps in Android:
    name=apps[0]
    reviews_no=float(apps[3])

    if name in reviews_max and reviews_max[name]<reviews_no:
        reviews_max[name]=reviews_no
    elif name not in reviews_max:
        reviews_max[name]=reviews_no



**In the previous cell we seen that total no. of duplicate entries are 1181. So length of our dictionary should be equal to difference of length of actual dataset and 1181.**

In [9]:
print("Actual length : ",len(Android)-1181)
print("\nExpected length : ",len(reviews_max))

Actual length :  9659

Expected length :  9659


**Now, let's use the reviews_max dictionary to remove the duplicates. For the duplicate cases, we'll only keep the entries with the highest number of reviews. In the code cell below:**

    .We start by initializing two empty lists, android_clean and already_added.
    .We loop through the android data set, and for every iteration:
              .We isolate the name of the app and the number of reviews.

**We add the current row (app) to the android_clean list, and the app name (name) to the already_cleaned list if:**

    .The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and
    .The name of the app is not already in the already_added list. 
    .We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). 
    .If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [10]:
Android_clean = []
Already_added = []

for apps in Android:
    name = apps[0]
    reviews_no = float(apps[3])
    
    if (reviews_max[name] == reviews_no) and (name not in Already_added):
        Android_clean.append(apps)
        Already_added.append(name) 
print(Android_clean[:5])

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']]


**Now explore the android_clean data set to ensure everything went as expected. The data set should have 9,659 rows.**

In [11]:
print(len(Android_clean),"\n")
explore_data(Android_clean , 0, 3,True)

9659 

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


No. of rows: 9659
No. of columns: 13


**We have got 9659 rows and 13 columns , just as expected.**

# Removing Non-English Apps:

## Part 1

If you explore the data sets enough, you'll notice the names of some of the apps suggest they are not directed toward an English-speaking audience. Below, we see a couple of examples from both data sets:



In [12]:
print(Ios[813][1])
print(Ios[6731][1])
print("\n")
print(Android_clean[4412][0])
print(Android_clean[7940][0])

爱奇艺PPS -《欢乐颂2》电视剧热播
【脱出ゲーム】絶対に最後までプレイしないで 〜謎解き＆ブロックパズル〜


中国語 AQリスニング
لعبة تقدر تربح DZ



In the previous step, we managed to remove the duplicate app entries in the Google Play data set. Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are directed toward an English-speaking audience. However, if we explore the data long enough, we'll find that both data sets have apps with names that suggest they are not directed toward an English-speaking audience.

**One way to go about this is to remove each app with a name containing a symbol that is not commonly used in English text — English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).**

Behind the scenes, each character we use in a string has a corresponding number associated with it. For instance, **the corresponding number for character 'a' is 97, character 'A' is 65, and character '爱' is 29,233.'** We can get the corresponding number of each character using the ord() built-in function.

We built this function below, and we use the built-in **ord()**  function to find out the corresponding encoding number of each character.

In [13]:
def  is_english(string):
    
    for character in string:
        if ord(character)>127:
            return False
        
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
    

True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

In [14]:
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(ord('😜'))
print(ord('™'))

False
False
128540
8482


# Part 2
**To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters:**

In [15]:
def is_english(string):
    non_ascii=0
    
    for character in string:
        if ord(character)>127:
            non_ascii+=1
    
    if non_ascii>3:
        return False
    else :
        return True
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
True


The function is still not perfect, and very few non-english apps might get past out filter, but this seems good enough at this point in our analysis - we shouldn't spend too much time on optimization at this point.

Below, we use the is_english() function to filter out the non_English apps for both data sets:

In [16]:
Android_english =[]
Ios_english=[]

for app in Android_clean:
    name=app[0]
    if  is_english(name):
        Android_english.append(app)
for app in Ios:
    name=app[1]
    if is_english(name):
        Ios_english.append(app)
explore_data(Android_english,0,3,True)
print('\n')
explore_data(Ios_english,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


No. of rows: 9614
No. of columns: 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games

We can see that we are left with **9614** Android apps and **6183** Ios apps.

## Isolating the Free Apps

As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-apps ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [17]:
Android_final=[]
Ios_final=[]
for app in Android_english:
    price = app[7]
    if price =='0':
        Android_final.append(app)

for app in Ios_english:
    price=app[4]
    if price=='0.0':
        Ios_final.append(app)

print(len(Android_final))
print(len(Ios_final))

8864
3222


Now We are left with 8864 Android Apps and 3222 Ios Apps, which should be enough for our analysis.

## Most Common Apps by Genre

### Part One

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

**To minimize risks and overhead our validation strategy for an app idea is comprised of three steps:**
    
    >>Build a minimal Android version of the app,and add it to Google play.
    >>If the app has a good response from users, we then develop it further.
    >>If the app is profitable after six months, we also build an Ios version of the app and add it to the App store.
 
Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successfull on both markets.For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres  for each market. For this we will build a frequency table for the prime_genre column of the App Store data set, and Genres and Catagory columns  of the Google Play data set.

### Part Two

**We will build two functions we can use to analyze the frequency tables:**
  
      >>One function to generates frequency tables that show percentages.
      >> Another function that we can use to display the percentages in a descending order.

In [18]:
def freq_table(dataset,index):
    table={}
    total=0
    
    for row in dataset:
        total+=1
        value=row[index]
        if value in table:
            table[value]+=1
        else :
            table[value]=1
            
    table_percentages={}
    for key in table:
        percentages=(table[key]/total)*100
        table_percentages[key]=percentages
    return table_percentages

def display_table(dataset,index):
    table=freq_table(dataset,index)
    table_display=[]
    for key in table:
        key_val_as_tuple=(table[key],key)
        table_display.append(key_val_as_tuple)
    table_sorted=sorted(table_display,reverse=True)
    for entry in table_sorted:
        print(entry[1],':',entry[0])

## Part Three
**We start by examining the frequency table for the prime_genre column of the App store data set.**

In [19]:
display_table(Ios_final,-5)

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


We can see that among the free English apps, more than a half (**58.1626%**) are games. Entertainment apps are close to **8%** ,followed by photo and video apps, which  are to **5%**. Only **3.66%** of the apps are designed for education, followed by Social Networking apps that are of **3.28988%** of the total apps of the data set.

The general impression is that App  Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyke, etc.) are more rare.
However, the fact that fun apps are the most numerous doesn't also imply that  they also have the greatest number of users - the demand might not be the same as the offer.

Let's continue by ecamining the Genres and Catagory columns of the Google Play  data set (two columns which seems to be related).

In [20]:
display_table(Android_final,1)

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and seems that a good number of apps are designed for practical purpuses (family, tools, business, lifestyle, productivity etc.). However, if we investigate this further, we can see that the family catagory (which is of almost **19%** of the apps) means mostly games for kids.

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:

In [21]:
display_table(Android_final,-4)

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

The  difference between the Genres and the Catagory columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (It has more catagories). we're only looking for the bigger picture at the moment, so we'll only work with the Catagory columnn moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for fun-apps,. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what  genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. For the Google Play  data set, we can find this information in the Installs column, but for the App Store data set this information is missing. As a workaround, we'll take total number of user ratings as a proxy, which we can find i n the rating_count_tot app.

**Below, we calculate the average number of user ratings per app genre on the App Store:**

In [22]:
genres_ios=freq_table(Ios_final,-5)
 
for genre in genres_ios:
    total=0
    len_genre=0
    for app in Ios_final:
        genre_app=app[-5]
        if genre_app == genre:
            n_ratings=float(app[5])
            total+=n_ratings
            len_genre +=1
    avg_n_ratings = total / len_genre
    print(genre,':',avg_n_ratings)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


On average, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [23]:
for app in Ios_final:
    if app[-5]=='Navigation':
        print(app[1],':',app[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc.
Same applies to music apps, where a few big players like Pandora, Spotify, Ganna, Saavan and Shazam heavily influence the average number.

Our aim is to fin popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 thresold. We could get a  better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps  have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [24]:
for app in Ios_final:
    if app[-5]=='Reference':
        print(app[1],':',app[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


However, this niche seems to show some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book, an audio version of book, quizzes about the book etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun  apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

Other genres seems popular include weather, book, food and drink, or finance. The book genre seem to overlap a bit with the app idea we described above, but the other genres don't seem too interesting to us:

    >>Weather apps- people generally don't spend too much time in-app, and the chances of making profit from in-app ads are low. Also, getting reliable live weather data may require us to connect our apps to non-free APIs.
    >>Food and drink - examples here include Starbucks, Dunkin' Donuts, McDonal's, Swiggy, Zomato etc. So making  a popular food and drink app requires actual cooking  and a delivery service, which is outside the scope of  our company.
    >>Finance apps  - these apps involve banking, paying bills , money transfer, etc. Building a finance app requires domain knowledge, and we don't want to hire a finance sepert just to build an app.
    
Now let's analyze  the Google Play market a bit.
 
## Most Popular Apps by Genre on Google Play
For  the Google Play market, we actually have data about the number of installs, so we should be able to get a  clearer picture about genre popularity. However, the install numbers don't seem precise enough - we can see that  most values are open-ended (100+, 1000+, 5000+, etc.):

In [25]:
display_table(Android_final,5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


One problem with  the data is thar is not precise. For instance, we son't whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes - we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the numbers of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with **100,000+** installs has **100,000** installs, and an app with **1,000,000+** installs has **1,000,000** installs and so on.

To perform computations, however we'll need to convert each  install number to float - this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (catagory).

In [26]:
catagories_android=freq_table(Android_final, 1)

for catagory in catagories_android:
    total=0
    len_catagory=0
    for app in Android_final:
        catagory_app=app[1]
        if catagory_app==catagory:
            n_installs=app[5]
            n_installs=n_installs.replace('+','')
            n_installs=n_installs.replace(',','')
            total+= float(n_installs)
            len_catagory+=1
    avg_n_installs = total/ len_catagory
    print(catagory,':',avg_n_installs)        

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

On average, communication apps have the most installs 38,456,119. This number is heavily skewed up by a few apps that have over one billion installs (WhatsApp, Facebook Messenger, Skype, Google Chrome, Gmail, and Hangouts), and a few others with over 100 and 500 million installs:


In [27]:
for app in Android_final:
    if app[1]== 'COMMUNICATION' and (app[5]=='1,000,000,000+' or app[5]=='500,000,000+' or app[5]=='100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

If we remove all the communication apps that have over 100 million installs, the average would be reduced roughly ten times.:

In [28]:
under_100_m=[]
for app in Android_final:
    n_installs=app[5]
    n_installs=n_installs.replace(',','')
    n_installs=n_installs.replace('+','')
    if  (app[1]=='COMMUNICATION') and (float(n_installs)< 100000000):
        under_100_m.append(float(n_installs))
sum(under_100_m) / len(under_100_m)
                           

3603485.3884615386

We see the same pattern for the video players catagory, which is the runner-up witth 24,727,872 installs. The market is dominated by apps like Youtube, Google Play Movies & TV, or MX Player The pattern is repeated for social apps (where we have giants like Facebook, Instagram, Google+, etc.), photography apps (Google Photos and other popular photo editors), or productivity apps (Microsoft Word, Dropbox, Google Calendar, Evernote, etc.).

Again, the main concern is that these app genres might seem more popular than they really are. Moreover, these niches seem to be dominated by a few giants who are hard to compete against.

The game genre seems pretty popular, but previosly we found out this part of the market seems a bit saturated, so we'd like to come up with a different app recommendation if possible.

 The books and refrence genre looks fairly popular as well, with an average number of installs of 8,767,811. It's interesting to explore this in more depth, since we found this genre that shows potential for being profitable on both the App Store and Google Play.
 
 Let's take a look at some of the apps from this genre and their number of installs:

In [29]:
for app in Android_final:
    if app[1]=='BOOKS_AND_REFERENCE':
        print(app[0],':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and refrence genre includes a variety of  apps: software for procrssing and reading ebooks, various collection of libraries, dictionaries, tutorials on programming language, etc. It seems there's still a  small number of extremely popular apps that skew the average.

In [30]:
for app in Android_final:
    if app[1]=='BOOKS_AND_REFERENCE' and (app[5]=='1,000,000,000+' or app[5]=='500,000,000+' or app[5]=='100,000,000+'):
        print(app[0],':',app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


However, it looks like there are only a few very popular apps , so this market still shows potential. Let's try to get some app ideas based on the kind of popularity (between 1,000,000 and 100,000,000 downloads):

In [31]:
for app in Android_final:
    if app[1]=='BOOKS_AND_REFERENCE' and ( app[5]=='1,000,000' or  app[5]=='5,000,000+' or app[5]=='10,000,000+' or app[5]=='50,000,000+'):
        print(app[0],':',app[5])

Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
FBReader: Favorite Book Reader : 10,000,000+
AlReader -any text book reader : 5,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Quran for Android : 10,000,000+
Dictionary.com: Find Definitions for English Words : 10,000,000+
English Dictionary - Offline : 10,000,000+
Bible KJV : 5,000,000+
NOOK: Read eBooks & Magazines : 10,000,000+
Dictionary : 10,000,000+
Spanish English Translator : 10,000,000+
Dictionary - Merriam-Webster : 10,000,000+
JW Library : 10,000,000+
Oxford Dictionary of English : Free : 10,000,000+
English Hindi Dictionary : 10,000,000+
English to Hindi Diction

This niche seems to be dominated by software for processing and reading ebooks, as well as various collections of libraries and dictionaries, so it's probably not a good idea to build similar apps since there'll be some significant competition.

We also notice there are quite a few apps built around the book **Quran**, which suggests that building an app around a popular book can be profitable, It seems that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and App Store markets.

However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book .This might include daily qoutes from the book, an audio version of the book , quizzes on the book, a forum where people can discuss the book, etc.

## Conclusions

In this project , we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Plsy and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book \, quizzes on the book, a forum where people can discuss the book, etc.


**--Vivek Pandey**

Refrences and thanks to Dataquest E-learning platform.